cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java

```diff
@@ -0,0 +1,361 @@
+
+    checkpoints.put(0, new MutableLogBufferCheckpoint(checkpoint.getOffset().getFileId(),
+                                                      checkpoint.getOffset().getFilePos(),
+                                                      checkpoint.getMaxEventTime()));
+    context.start();
+    LOG.info("Log processor pipeline for {} with config {} started with checkpoint {}", name, config, this.checkpoints);
+  }
+
+  @Override
+  protected void run() throws Exception {
+    runThread = Thread.currentThread();
+
+    try {
+      lastCheckpointTime = System.currentTimeMillis();
+      while (!stopped) {
+        boolean hasEventProcessed = processEvents(incomingEventQueue);
+        long now = System.currentTimeMillis();
+        long nextCheckpointDelay = trySyncAndPersistCheckpoints(now);
+
+        // If nothing has been processed (e.g. fail to append anything to appender),
+        // Sleep until min(next checkpoint delay, next event delay).
+        if (!hasEventProcessed) {
+          long sleepMillis = config.getEventDelayMillis();
+          sleepMillis = Math.min(sleepMillis, nextCheckpointDelay);
+          if (sleepMillis > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepMillis);
+          }
+        }
+      }
+    } catch (InterruptedException e) {
+      // Interruption means stopping the service.
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    stopped = true;
+    if (runThread != null) {
+      runThread.interrupt();
+    }
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.debug("Shutting down log processor pipeline for {}", name);
+
+    try {
+      context.stop();
+      // Persist the checkpoints. It can only be done after successfully stopping the appenders.
+      // Since persistCheckpoint never throw, putting it inside try is ok.
+      persistCheckpoints();
+    } catch (Exception e) {
+      // Just log, not to fail the shutdown
+      LOG.warn("Exception raised when stopping pipeline {}", name, e);
+    }
+
+    LOG.info("Log processor pipeline for {} stopped with latest checkpoint {}", name, checkpoints);
+  }
+
+  @Override
+  protected String getServiceName() {
+    return "LogPipeline-" + name;
+  }
+
+  /**
+   * Reads events from incomingEventQueue and sends them to event processor for further processing.
+   */
+  private boolean processEvents(BlockingQueue<LogBufferEvent> incomingEventQueue) {
+    ProcessedEventMetadata<LogBufferFileOffset> metadata
+      = eventQueueProcessor.process(0, new LogFileOffsetTransformIterator(incomingEventQueue));
+
+    // none of the events were processed.
+    if (metadata.getTotalEventsProcessed() <= 0) {
+      return false;
+    }
+
+    unSyncedEvents += metadata.getTotalEventsProcessed();
+    // events were processed, so update the checkpoints
+    Checkpoint<LogBufferFileOffset> checkpoint = metadata.getCheckpoints().get(0);
+    MutableLogBufferCheckpoint mutableCheckpoint = checkpoints.get(0);
```

Comment:
I'm a bit confused about this map, there is only ever one item in it right? Is it used to make it conform to the checkpointManager API? Are the keys supposed to be partition numbers or something like that?

In any case, I think it's more clear if we also make the checkpoint a variable, and just use the map when passing it to the checkpoint manager.